### PR TITLE
ADR025: plan naming conventions

### PR DESCRIPTION
**Pull request by chrisfarms**
# What

We want to be consistent with the naming of our service plans and make
it easier to introduce new plan names for different versions.

This ADR was the result of a [review of plan names](https://trello.com/c/mdu7HdyE/9-review-our-service-plans ) and a [proposal](https://docs.google.com/document/d/10f1_6Uzg6VcS60luYLK61Xa54-wpMhz8K-VyZwJzhw0/)

# How to review

Make sure it makes sense, renders correctly

# Who can review

Not @chrisfarms

### CODE REVIEW

File: source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+tiny-clustered => tiny-clustered-3.2
+tiny-unclustered => tiny-unclustered-3.2
+```
+
+#### Additionally:
+
+* We will not use the word "free" in any names.
+* We will drop redundent words (like 'dedicated') from names to reduce noise.
+* We will use lowercase letters for names.
+* We will not use abbriviations for sizes ("medium" instead of "M", "large" instead of "L")
```
[the remark this answers is not in the document]
Comment (chrisfarms) [reply source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb]: You can't totally do away with the "size" since this is the part that defines the instance class.

I think we should stay away from placing storage sizes for RDS databases in plan names. It's possible that we might want to allow setting the storage sizes via update parameters to decouple the storage from the plan "size" which really refers to the VM instance (memory) size. The cost of the storage is almost negligible compared to the cost of the VM.

Since the "sizes" generally refer to VM instance classes which we know ahead of time it's less likely we will need to "add one in between".

A similar issue is there for redis with number of nodes. We would not want to have to offer `tiny-unclusted-1node`, `tiny-unclusted-2node`, `tiny-unclusted-3node` .... and then `medium-unclusted-1node`, `medium-unclusted-2node` .... when we could instead offer a `tiny-unclustered` and `medium-unclustered` and allow setting the number of nodes as a configurable parameter.

So essentially plan names are like "base configurations" of the big important things that are hard to switch between and things like storage or scaling etc can be tweaks on top of that if required. We know billing can be extended to work this way, because that's what we have had to do to handle compose's auto-scaling. We haven't done much around configuring individual instances mainly because the UI of writing json on the cli is nasty ... but now we have pazmin, which could open doors for a nicer UI for service configuration.

Basically that was a long way of saying. I think we need to stick with the `{size}` for now :)

File: source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+* We will only support encrypted RDS plans going forward. Legacy unencrypted plans to be explicitly labelled with -unencrypted and hidden from the marketplace.
+
+
+## Status
+
+Accepted
+
+## Consequences
+
+Renaming exisiting plans to conform to this format may be disruptive to tenants who may rely on the old names in scripted deployements.
```
[the remark this answers is not in the document]
Comment (chrisfarms) [reply source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb]: We've agreed to send out a mail to tenants to give them a few weeks heads up before the change.

File: source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#### For example:
+
+A large multi-az postgres plan for version 9.6 would be `large-ha-9.6`.
+
+A small not multi-az, unclustered, redis 3.2 plan would be `redis-unclustered-3.2`.
+
+#### Example migrations of some existing plans:
+
+```
+L-HA-enc-dedicated-9.5 => large-ha-9.5
```
[the remark this answers is not in the document]
Comment (chrisfarms) [reply source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb]: I think this will be very difficult to avoid clashes since we cannot really predict which versions are important ahead of time or even if versions will be a semver style at all. Some weird service might just use codenames for versions or something.

Versions "10.9.9" and "1.0.99" would both map to "1099" under that scheme ... which means if it ever comes up we would have to make an exception and call it something else. So I think it's better to just use the version string verbatim and avoid clashes.

**Comment by chrisfarms** [comment]
Renamed to ADR025 rather than continuing after the ADR444 anomaly

**Comment by keymon** [comment]
LGTM

File: source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+Adding additional versions of services has the potential to compound the
+problem by multiplying plans of different characteristics with different
+versions.
+
+## Decision
+
+We have decided to use the following naming convention for naming plans going forward:
+
+```
+{SIZE}-{HA}-{LABEL}-{LABEL}-{VERSION}
```
[the remark this answers is not in the document]
Comment (samcrang) [reply source/architecture_decision_records/ADR445-service-plan-naming-conventions.html.md.erb]: If HA is optional, isn't it just another type of label?
